File: tasks/python/task-07-inventory-negative-price/solution/inventory.py
"""Track stock for a small shop.

The inventory is a dict. The key is the SKU. The value is a dict with
a price and a quantity.
"""


def add_item(inventory, sku, price, quantity=0):
    """Add a new item.

    Raise ValueError if the SKU already exists or the price is negative.
    """
    if sku in inventory:
        raise ValueError(f"duplicate sku: {sku}")
    if price < 0:
        raise ValueError(f"negative price: {price}")
    inventory[sku] = {"price": price, "quantity": quantity}


def restock(inventory, sku, amount):
    """Add `amount` units to an existing item."""
    inventory[sku]["quantity"] += amount


def sell(inventory, sku, amount):
    """Remove `amount` units. Raise ValueError if there is not enough stock."""
    item = inventory[sku]
    if amount > item["quantity"]:
        raise ValueError(f"not enough stock for {sku}")
    item["quantity"] -= amount


def stock_value(inventory):
    """Return the value of all stock at the current prices."""
    return sum(item["price"] * item["quantity"] for item in inventory.values())


def out_of_stock(inventory):
    """Return the SKUs with zero units, in insertion order."""
    return [sku for sku, item in inventory.items() if item["quantity"] == 0]

File: tasks/typescript/task-07-inventory-negative-price/solution/inventory.ts
/**
 * Track stock for a small shop.
 *
 * The inventory is a Map. The key is the SKU. The value holds
 * a price and a quantity.
 */

export type StockItem = {
  price: number;
  quantity: number;
};

export type Inventory = Map<string, StockItem>;

/**
 * Add a new item.
 *
 * Throw an Error if the SKU already exists. Throw a RangeError if the price is negative.
 */
export function addItem(inventory: Inventory, sku: string, price: number, quantity = 0): void {
  if (inventory.has(sku)) {
    throw new Error(`duplicate sku: ${sku}`);
  }
  if (price < 0) {
    throw new RangeError(`negative price: ${price}`);
  }
  inventory.set(sku, { price, quantity });
}

/** Return the item for a SKU. Throw an Error if the SKU is unknown. */
function getItem(inventory: Inventory, sku: string): StockItem {
  const item = inventory.get(sku);
  if (item === undefined) {
    throw new Error(`unknown sku: ${sku}`);
  }
  return item;
}

/** Add `amount` units to an existing item. */
export function restock(inventory: Inventory, sku: string, amount: number): void {
  getItem(inventory, sku).quantity += amount;
}

/** Remove `amount` units. Throw a RangeError if there is not enough stock. */
export function sell(inventory: Inventory, sku: string, amount: number): void {
  const item = getItem(inventory, sku);
  if (amount > item.quantity) {
    throw new RangeError(`not enough stock for ${sku}`);
  }
  item.quantity -= amount;
}

/** Return the value of all stock at the current prices. */
export function stockValue(inventory: Inventory): number {
  let value = 0;
  for (const item of inventory.values()) {
    value += item.price * item.quantity;
  }
  return value;
}

/** Return the SKUs with zero units, in insertion order. */
export function outOfStock(inventory: Inventory): string[] {
  const skus: string[] = [];
  for (const [sku, item] of inventory) {
    if (item.quantity === 0) {
      skus.push(sku);
    }
  }
  return skus;
}
